Modules/_functoolsmodule.c

```diff
@@ -41,13 +42,94 @@ get_functools_state(PyObject *module)
 
 /* partial object **********************************************************/
 
+
+/*
+Placeholder is an object that can be used to signal that positional
+   argument place is empty when using `partial` class
+*/
+
+#define PY_CONSTANT_PLACEHOLDER 0
+
+static PyObject* constants[] = {
+    NULL,         // PY_CONSTANT_PLACEHOLDER
+};
+
+typedef struct {
+    PyObject_HEAD
+} placeholderobject;
+
+PyDoc_STRVAR(placeholder_doc,
+"PlaceholderType()\n"
+"--\n\n"
+"The type of the Placeholder singleton.\n"
+"Used as a placeholder for partial arguments.");
+
+static PyObject *
+placeholder_repr(PyObject *op)
+{
+    return PyUnicode_FromString("Placeholder");
+}
+
+static void
+placeholder_dealloc(PyObject* placeholder)
+{
+    /* This should never get called, but we also don't want to SEGV if
+     * we accidentally decref Placeholder out of existence. Instead,
+     * since Placeholder is an immortal object, re-set the reference count.
+     */
+    _Py_SetImmortal(placeholder);
+}
+
+static PyObject *
+placeholder_new(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+    if (PyTuple_GET_SIZE(args) || (kwargs && PyDict_GET_SIZE(kwargs))) {
+        PyErr_SetString(PyExc_TypeError, "PlaceholderType takes no arguments");
+        return NULL;
+    }
+    PyObject *singleton = constants[PY_CONSTANT_PLACEHOLDER];
+    if (singleton == NULL) {
+        singleton = PyType_GenericNew(type, NULL, NULL);
+        constants[PY_CONSTANT_PLACEHOLDER] = singleton;
+    }
+    return singleton;
+}
+
+static int
+placeholder_bool(PyObject *v)
+{
+    PyErr_SetString(PyExc_TypeError,
+                    "Placeholder should not be used in a boolean context");
+    return -1;
+}
+
+static PyType_Slot placeholder_type_slots[] = {
+    {Py_tp_dealloc, placeholder_dealloc},
+    {Py_tp_repr, placeholder_repr},
+    {Py_tp_doc, (void *)placeholder_doc},
+    {Py_tp_new, placeholder_new},
+    // Number protocol
+    {Py_nb_bool, placeholder_bool},
+    {0, 0}
+};
+
+static PyType_Spec placeholder_type_spec = {
+    .name = "functools.PlaceholderType",
```

Comment:
```suggestion
    .name = "functools._PlaceholderType",
```

I don't want to have a public type honestly.